### PR TITLE
HTTP client improvements

### LINKED ISSUE #3500: HTTP client eager connect
The HTTP client design creates HTTP requests before they are actually connected which means they can be used before a request is connected or obtained from the pool. This design forces to buffer request data before leading to extra complexity in the code and race bugs.

Removing this might seem like removing some usability in the client, the impact is fine because most users seeking for usability are using the Web client that provides a different model where the stream is actually managed by the client.

### PR DESCRIPTION
**Pull request by vietj**
fixes #3500 and #3501

### CODE REVIEW

**Comment by vietj** [comment]
Note this PR will require to update a few projects in the stack because of API changes

File: src/main/java/io/vertx/core/http/impl/HttpClientPush.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (c) 2011-2019 Contributors to the Eclipse Foundation
```
Comment (jponge) [inline src/main/java/io/vertx/core/http/impl/HttpClientPush.java:2]: 2020 😉

File: src/test/java/io/vertx/core/spi/tracing/HttpTracerTestBase.java [hunk 2]
```diff
@@ -156,8 +160,8 @@ public Object sendRequest(Context context, Object request, String operation, BiC
       public void receiveResponse(Context context, Object response, Object payload, Throwable failure, TagExtractor tagExtractor) {
         assertSame(val, context.getLocal(key));
         assertTrue(context.removeLocal(key));
-        assertNotNull(response);
-        assertTrue(response instanceof HttpClientResponse);
+//        assertNotNull(response);
```
Comment (tsegismont) [inline src/test/java/io/vertx/core/spi/tracing/HttpTracerTestBase.java:163]: Update or remove?

File: src/main/java/io/vertx/core/http/impl/VertxHttp2Stream.java [hunk 3]
```diff
@@ -210,10 +212,20 @@ final void writeData(ByteBuf chunk, boolean end, Handler<AsyncResult<Void>> hand
...
+      chunk = Unpooled.EMPTY_BUFFER;
+    } else {
+      chunk = buf;
+    }
+    if (chunk != null) {
+      bytesWritten += chunk.readableBytes();
+      FutureListener<Void> promise = handler == null ? null : context.promise(handler);
+      conn.handler.writeData(stream, chunk, end, promise);
+    } else {
+      // todo ??????????
```
Comment (tsegismont) [inline src/main/java/io/vertx/core/http/impl/VertxHttp2Stream.java:227]: todo, fail ?

**Comment by vietj** [comment]
I think this is good for merging now - I will merge this soon and update the rest of the stack with fixes.

**Pull request merged**